internal/hardware/hardware.go

```diff
@@ -108,3 +125,83 @@ func GetMutableHardwareInfoDelta(hardwareMutableInfoPrevious models.HardwareInfo
 
 	return hardwareInfo
 }
+
+// GetDevices recursively traverses a directory specified by path
+// and returns all devices found there.
+// Shamelessly copied from 	https://github.com/opencontainers/runc/blob/main/libcontainer/devices/device_unix.go
+func getDevices(path string) ([]*models.HostDevice, error) {
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+	var out []*models.HostDevice
+	for _, f := range files {
+		switch {
+		case f.IsDir():
+			switch f.Name() {
+			// ".lxc" & ".lxd-mounts" added to address https://github.com/lxc/lxd/issues/2825
+			// ".udev" added to address https://github.com/opencontainers/runc/issues/2093
+			case "pts", "shm", "fd", "mqueue", ".lxc", ".lxd-mounts", ".udev":
+				continue
+			default:
+				sub, err := getDevices(filepath.Join(path, f.Name()))
```

Comment:
is it safe from cyclic paths references? (an endless recursion that will lead to undesired result)

Reply:
no. I don't think so.